File: java/RouteMap.java
package irresponsible.crouton;

import clojure.lang.IFn;
import clojure.java.api.Clojure;
import java.lang.IllegalArgumentException;
import java.util.HashMap;
import java.util.Map;
import clojure.lang.IPersistentVector;
import clojure.lang.ITransientMap;
import clojure.lang.RT;

public class RouteMap implements IRoute {
  private final HashMap<String, IRoute> map;
  public RouteMap(Map<String, IRoute> m) {
    if (null == m)
      throw new IllegalArgumentException("map must not be nil");
    map = new HashMap<String,IRoute>(m);
  }
  public final Object match(IPersistentVector pieces, ITransientMap places) {
    String piece = (String) pieces.nth(0, null);
    if (piece == null) return null;
    IRoute next = map.get(piece);
    return (next != null) ? next.match(RT.subvec(pieces, 1, pieces.count()), places) : null;
  }
}

File: java/RouteMap.kt
package irresponsible.crouton

import java.util.HashMap
import clojure.lang.IPersistentVector
import clojure.lang.IPersistentMap
import clojure.lang.RT

class RouteMap(m: Map<String, IRoute>?) : IRoute {
    private val map: HashMap<String, IRoute>

    init {
        if (null == m)
            throw IllegalArgumentException("map must not be nil")
        map = HashMap(m)
    }

    override fun match(pieces: IPersistentVector, places: IPersistentMap): Any? {
        val piece = pieces.nth(0, null) as String ?: return null
        val next = map[piece]
        return next?.match(RT.subvec(pieces, 1, pieces.count()), places)
    }
}
